### DISCUSSION TITLE
Interactive documenation

### DISCUSSION THREAD

**Discussion by viebel** [josdejong/mathjs#2715] [Ideas]
Mathjs really rocks!

It would be great to have interactive documentation for math.js.

It's super easy to create interactive documenation with the [klipse plugin](github.com/viebel/klipse).

As a POC, I've written a short interactive article that demonstrate the power of `math.js` evaluator:
http://read.klipse.tech/symbolic-calculation-in-javacript/

Disclaimer: I'm the author of the [klipse plugin](github.com/viebel/klipse).

**Comment by josdejong** [comment]
Thanks. Nice idea to make the code examples in the documentation interactive. 

One solution is to embed an editor, like your klipse plugin or what I did with Math Notepad ([see these docs](http://mathnotepad.com/docs/overview.html), you can edit all code).

Another solution is to put a link on top or at the bottom which copies the code to say http://jsbin.com (or a klipse web service) so the user can edit and save everything himself (like the "click here to fiddle around" on the front page of http://mathjs.org/). I may like this approach more since it gives the user more control and freedom to play around. What do you think?

**Comment by FSMaxB** [comment]
I think a link to something like jsbin should be in addition to the inline editor, because it pulls people out of context.

**Comment by viebel** [comment]
I agree with @fsmaxb 
The best reading experience to me is interactive and without having to leave the page. 

Would you like to try integrating klipse on your doc pages?

**Comment by josdejong** [comment]
Thanks Max, that's a good point.

Yes Yehonathan, sounds good to give this a try integrating klipse in the website of math.js. I haven't thought out in details what would be needed there. There is a script which generates the html docs from markdown, I suppose we have to add some code there to embed klipse where applicable (maybe via an annotation in the markdown code before or around the relevant code blocks, or maybe we can simply match block fences tagged with `js`). Would you like to give this a try @viebel and see whether you can get klipse integrated?

**Comment by viebel** [comment]
What is the script that generates html?
Is there a way to add a script tag to all the pages?

**Comment by josdejong** [comment]
It's all in the `gh-pages` branch. There is a gulpfile with a couple of tasks. The most relevant part to get started is the `'docs'` task, which does some preprocesses of the markdown docs:

https://github.com/josdejong/mathjs/blob/gh-pages/gulpfile.js#L102-L114

The actual generation of HTML from the markdown is done by Github Pages (which uses jekyll), when you push something new to `gh-pages` and the site is deployed.

**Comment by josdejong** [comment]
The readme in the root of `gh-pages` explains how to built the website locally:

https://github.com/josdejong/mathjs/blob/gh-pages/README.md

**Comment by viebel** [comment]
Here is a 1st draft #771

**Comment by viebel** [comment]
@josdejong did you have time to review #771  ?

**Comment by josdejong** [comment]
Not yet, will look at it asap (work, family, stuff, you know...), I expect to be able to check it out tomorrow night.

**Comment by viebel** [comment]
No hurry. Take your time...
